iml-agent/src/action_plugins/check_ha.rs

```diff
@@ -71,27 +92,208 @@ impl AgentInfo {
+            _ => false,
+        },
+    )
+}
+
+/// property: org.freedesktop.systemd1
+/// format!("/org/freedesktop/systemd1/unit/{}_2eservice", name)
+/// UnitFileState   +   ActiveState  -> RunState
+/// disabled            inactive        Stopped
+/// enabled             inactive        Enabled
+/// disabled            active          Started
+/// enabled             active          Setup
+async fn systemd_unit_props(name: &str) -> Result<RunState, ImlAgentError> {
+    let s = format!("/org/freedesktop/systemd1/unit/{}_2eservice", name);
+    cmd_output(
+        "busctl",
+        &[
+            "--system",
+            "--no-pager",
+            "get-property",
+            "org.freedesktop.systemd1",
+            &s,
+            "org.freedesktop.systemd1.Unit",
+            "UnitFileState",
+            "ActiveState",
+        ],
+    )
+    .compat()
+    .await
+    .and_then(|output| {
+        let res: Vec<&str> = str::from_utf8(&output.stdout)
+            .unwrap()
+            .lines()
+            .map(|o| busctl_to_string(o))
+            .collect();
+        match res[..] {
+            ["disabled", "inactive"] => Ok(RunState::Stopped),
+            ["enabled", "inactive"] => Ok(RunState::Enabled),
+            ["disabled", "active"] => Ok(RunState::Started),
+            ["enabled", "active"] => Ok(RunState::Setup),
+            _ => Err(ImlAgentError::from(RequiredError(format!(
+                "Unknown busctl ({}) output: {:?}",
+                s, output.stdout
+            )))),
+        }
+    })
+}
+
+async fn systemd_unit_servicestate(name: &str) -> Result<ServiceState, ImlAgentError> {
+    if systemd_unit_enabled(name).await? {
+        Ok(ServiceState::Configured(systemd_unit_props(name).await?))
+    } else {
+        Ok(ServiceState::Unconfigured)
+    }
+}
+
+async fn file_exists(path: &str) -> bool {
+    match cmd_output("test", &["-f", path]).compat().await {
+        Err(_) => false,
+        Ok(o) => o.status.success(),
+    }
+}
+
+/// systemctl is-enabled SERVICE :: call: org.freedesktop.systemd1
+/// /org/freedesktop/systemd1
+/// org.freedesktop.systemd1.Manager.GetUnitFileState SERVICE
+///
+/// systemctl is-running SERVICE :: property: org.freedesktop.systemd1
+/// format!("/org/freedesktop/systemd1/unit/{}_2eservice", name)
+/// [org.freedesktop.systemd1.Unit] UnitFileState + ActiveState
+/// corosync    - service == Unconfigured if ! -f /etc/corosync/corosync.conf
+/// pacemaker   - service == Unconfigured if ! -f /var/lib/pacemaker/cib/cib.xml
+/// pcsd        - service == Unconfigured if ! -f /var/lib/pcsd/tokens
+
+async fn check_corosync() -> Result<ComponentState<bool>, ImlAgentError> {
+    let mut corosync = ComponentState::<bool> {
+        ..Default::default()
+    };
+
+    if file_exists("/etc/corosync/corosync.conf").await {
+        corosync.config = ConfigState::Unknown;
+        corosync.service = systemd_unit_servicestate("corosync").await?;
+        // @@ check corosync setup
```

Comment:
Is this a "Todo" and if so can it be done in a separate patch or should it be done in this one?

Reply:
Looks like this is still a todo?

Reply:
Just following up on this. I'm assuming there is still some checks that need to be added?